### PR TITLE
fix: emit eval wrapper into target's package when this is non-public (#7)

### LINKED ISSUE #7: Expression evaluator cannot reach package-private fields on a non-public `this`
## Symptom
Reported by an agent driving the jdwp-inspector plugin:
> Expression compiler can't reference package-private classes' fields (e.g., `this.referencedEntityToGroupIdTranslator` on a package-private `BitmapSlicer`) — neither directly nor via reflection (unhandled `NoSuchFieldException` / `IllegalAccessException`).

## Root cause
`JdiExpressionEvaluator` always emits the generated wrapper class into the isolated package `mcp.jdi.evaluation`. From there, package-private members of the target type are unreachable, and the auto-rewrite of `this.field` → `_this.field` (`JdiExpressionEvaluator.java:546-553`) is gated on `thisClass.isPublic()` AND `Field::isPublic`, so any non-public `this` short-circuits the rewrite entirely.

The wrapper class lives in a fixed package, so even if the rewriter were widened, the field access would still be rejected at compile time.

## Proposal (option A from chat)
For a non-public `this` in an addressable package (i.e. not `java.*` / `javax.*` / `sun.*` / `jdk.*`, not a local/anonymous class), emit the wrapper into `this`'s own package. Same module + same package = package-private members become reachable.

- New `resolveWrapperPackage(thisObject)` picks the wrapper's package.
- `getDeclaredType(ReferenceType, wrapperPackage)` returns the type's actual name (not a public supertype) when it lives in the wrapper's package; binary `\$` is rewritten to source `.` for nested classes; `\$\$` proxy markers preserved.
- The field-rewrite filter widens to include public/protected/package-private fields when the wrapper shares `this`'s package. Private fields still aren't rewritten.

## Out of scope
- Reflection-based fallback for `private` fields (separate concern).
- Same trick for `static` methods on non-public declaring classes (worth a follow-up but not in this issue).

### PR DESCRIPTION
**Pull request by novoj**
Closes #7.

## Summary
Generated wrapper classes used by the expression evaluator now live in **the target type's own package** when `this` is non-public, so package-private fields and non-public types become reachable instead of being short-circuited to a public supertype.

## How
- `resolveWrapperPackage(thisObject)` decides whether to emit into `this`'s package (non-public + addressable) or fall back to the isolated `mcp.jdi.evaluation`.
- `getDeclaredType(ReferenceType, wrapperPackage)` now treats same-package non-public types as reachable, returning their real name (with `$` → `.` for nested classes, `$$` proxy markers preserved).
- The auto-rewrite of `this.field` → `_this.field` widens to include public/protected/package-private fields in shared-package mode. Private fields stay un-rewritten.
- Restricted packages (`java.*`, `javax.*`, `sun.*`, `jdk.*`) and local/anonymous classes route back to the default isolated package.

## Test plan
- [x] `./mvnw -pl jdwp-mcp-server test` — 904 tests pass (4 new in `JdiExpressionEvaluatorGetDeclaredTypeTest`)
- [x] Compile clean — no new NullAway / Error Prone warnings
- [ ] Test flight against a target VM with a package-private `this` to confirm the wrapper compiles + executes in the target module (defineClass into a named module is environment-dependent — works for non-modular Spring Boot apps, may need follow-up for fully modularized targets).

## Risk
The big unknown is **modular Java apps**: `ClassLoader.defineClass(name, bytes, 0, len)` places the wrapper into the classloader's unnamed module unless the target package is registered with a named module. In modular targets where `com.example` is a named module, the wrapper may end up in the unnamed module and lose package-private access. The fallback is the existing isolated-package behavior + a public-supertype walk, so the worst case is "same compile error as before".

### CODE REVIEW

File: jdwp-mcp-server/src/main/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluator.java [hunk 1]
```diff
@@ -423,20 +445,128 @@ private static String getDeclaredType(ReferenceType type) {
...
+     * NOT nested-class separators and rewriting them produces nonsense like {@code Bar..Mock}.
+     * Local and anonymous classes ({@code $<digit>}) must be filtered upstream; this method
+     * does not reject them because the same conversion is occasionally useful for diagnostics.
+     */
+    private static String toSourceTypeName(String binaryName) {
+        if (binaryName.indexOf('$') < 0 || binaryName.contains("$$")) {
+            return binaryName;
+        }
+        return binaryName.replace('$', '.');
+    }
```
[the remark this answers is not in the document]
Comment (novoj) [reply jdwp-mcp-server/src/main/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluator.java:649]: Good catch — fixed in 1c72c09. `toSourceTypeName` now only rewrites `$` when it sits between two non-`$` Java identifier-part chars AND the preceding char is not `.`. So `com.sun.proxy.$Proxy12` passes through unchanged. Added a regression test `shouldNotRewriteJdkDynamicProxyName`.

File: jdwp-mcp-server/src/main/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluator.java [hunk 2]
```diff
@@ -423,20 +445,128 @@ private static String getDeclaredType(ReferenceType type) {
...
+     */
+    private static boolean isReachableFromWrapper(ClassType type, String wrapperPackage) {
+        if (type.isPublic()) {
+            return true;
+        }
+        if (!packageOf(type.name()).equals(wrapperPackage)) {
+            return false;
+        }
+        return !isLocalOrAnonymous(type.name());
+    }
```
[the remark this answers is not in the document]
Comment (novoj) [reply jdwp-mcp-server/src/main/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluator.java]: Agreed and fixed in 1c72c09. `isReachableFromWrapper` now short-circuits on `type.isPrivate()` BEFORE the same-package check, so a private nested class walks to its public supertype instead of leaking through. Added test `shouldNotExposePrivateNestedClassEvenInSamePackage`.

File: jdwp-mcp-server/src/main/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluator.java [hunk 3]
```diff
@@ -534,24 +664,36 @@ private static ClassLoaderReference findClassLoader(StackFrame frame) throws Jdi
...
+            //    and produce a compile error the user can fix (or work around via reflection).
+            if (thisObject != null && thisObject.referenceType() instanceof ClassType thisClass) {
+                final boolean sharingPackage = !wrapperPackage.equals(DEFAULT_EVALUATION_PACKAGE)
+                    && wrapperPackage.equals(packageOf(thisClass.name()));
+                if (thisClass.isPublic() || sharingPackage) {
+                    final Set<String> shadowingLocals = context.getVariables().stream()
+                        .map(v -> v.name)
+                        .collect(Collectors.toSet());
+                    final Set<String> rewritableFieldNames = thisClass.allFields().stream()
+                        .filter(f -> sharingPackage ? !f.isPrivate() : f.isPublic())
```
[the remark this answers is not in the document]
Comment (novoj) [reply jdwp-mcp-server/src/main/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluator.java]: Yes, valid concern — fixed in 1c72c09. Extracted the per-field accessibility check into a new helper `isFieldAccessibleFromWrapper(field, wrapperPackage, sharingPackage)` which now ALSO requires `packageOf(field.declaringType().name()) == wrapperPackage` for non-public fields. A protected field inherited from a class in another package stays un-rewritten (the wrapper is not a subclass).

File: jdwp-mcp-server/src/main/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluator.java [hunk 4]
```diff
@@ -329,13 +337,13 @@ private static EvaluationContext buildContext(StackFrame frame, Map<String, Valu
...
-     * {@code java.lang.Object} for {@code null} and unrecognised value kinds.
+     * walk performed by {@link #getDeclaredType(ReferenceType, String)} so a non-public exception
+     * subclass still binds as a public ancestor (worst case {@code java.lang.Object}). Falls back
+     * to {@code java.lang.Object} for {@code null} and unrecognised value kinds.
      */
-    private static String inferDeclaredType(@Nullable Value value) {
+    private static String inferDeclaredType(@Nullable Value value, String wrapperPackage) {
         if (value instanceof ObjectReference objRef) {
-            return getDeclaredType(objRef.referenceType());
+            return getDeclaredType(objRef.referenceType(), wrapperPackage);
```
[the remark this answers is not in the document]
Comment (novoj) [reply jdwp-mcp-server/src/main/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluator.java:351]: You were right — there was a stale `{@link JdiExpressionEvaluator#inferDeclaredType(Value)}` in `JdiEventListener.java:796`. Fixed in 1c72c09 by switching to `{@code …}` (the method is private and the reference is purely informational, so a hard `@link` was overkill anyway).

File: jdwp-mcp-server/src/test/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluatorGetDeclaredTypeTest.java [hunk 5]
```diff
@@ -16,13 +16,20 @@
...
+ * {@code getDeclaredType(ReferenceType, String)} helper. This routine resolves the type name
+ * the wrapper class should reference, taking care of dynamic proxies (CGLIB / Guice / Mockito
+ * {@code $$}) and non-public types (walks up to a public supertype, or to a same-package type
+ * the wrapper can see by being emitted into that package, falling back to {@code java.lang.Object}).
  */
 class JdiExpressionEvaluatorGetDeclaredTypeTest {
 
+	/**
+	 * The default isolated wrapper package — passing this disables the "share package with target"
+	 * shortcut and exercises the legacy walk-to-public-supertype behaviour.
```
[the remark this answers is not in the document]
Comment (novoj) [reply jdwp-mcp-server/src/test/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluatorGetDeclaredTypeTest.java]: Good point — reworded in 1c72c09 to "exercises the legacy walk-to-public-supertype behaviour for typical target types" with an explicit note that a non-public type whose package equals `mcp.jdi.evaluation` would still trigger the shortcut (theoretical edge case not exercised by the tests).

File: jdwp-mcp-server/src/main/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluator.java [hunk 6]
```diff
@@ -534,24 +664,36 @@ private static ClassLoaderReference findClassLoader(StackFrame frame) throws Jdi
...
-                    .map(v -> v.name)
-                    .collect(Collectors.toSet());
-                final Set<String> publicFieldNames = thisClass.allFields().stream()
-                    .filter(Field::isPublic)
-                    .map(Field::name)
-                    .collect(Collectors.toSet());
-                expression = rewriteThisFieldReferences(expression, publicFieldNames, shadowingLocals);
+            // The set of rewritable fields depends on what the wrapper can actually see:
+            //  - When the wrapper lives in the default isolated package, only public fields on a
+            //    public declaring type are reachable.
```
[the remark this answers is not in the document]
Comment (novoj) [reply jdwp-mcp-server/src/main/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluator.java]: Fixed in 1c72c09. The `evaluate()` javadoc now describes BOTH branches of the auto-rewrite — public `this` OR wrapper sharing `this`'s package — and references `isFieldAccessibleFromWrapper` for the per-field accessibility check.

File: jdwp-mcp-server/src/main/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluator.java [hunk 7]
```diff
@@ -423,20 +445,197 @@ private static String getDeclaredType(ReferenceType type) {
...
+        if (isLocalOrAnonymous(binaryName)) {
+            return DEFAULT_EVALUATION_PACKAGE;
+        }
+        final String pkg = packageOf(binaryName);
+        // Refuse to define classes into restricted JDK packages — `java.*` is enforced by the JVM
+        // itself, the rest are flagged here so we get a clean log message rather than a runtime
+        // SecurityException far from the decision site.
+        if (pkg.isEmpty() || pkg.startsWith("java.") || pkg.equals("java")
+            || pkg.startsWith("javax.") || pkg.equals("javax")
+            || pkg.startsWith("sun.") || pkg.startsWith("jdk.")) {
```
[the remark this answers is not in the document]
Comment (novoj) [reply jdwp-mcp-server/src/main/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluator.java]: Good catch — fixed in f6a0c2d. `resolveWrapperPackage` now also excludes the bare `sun` and `jdk` package names (`pkg.equals("sun")` / `pkg.equals("jdk")`) alongside the existing prefix checks, mirroring the symmetric `java`/`javax` handling. A class declared directly in package `sun` or `jdk` now falls back to the default `mcp.jdi.evaluation` wrapper package instead of attempting a restricted injection.

File: jdwp-mcp-server/src/main/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluator.java [hunk 8]
```diff
@@ -423,20 +445,198 @@ private static String getDeclaredType(ReferenceType type) {
...
                 }
                 if (current != null) {
-                    return current.name();
+                    return toSourceTypeName(current.name());
                 }
                 return "java.lang.Object";
             }
 
-            return typeName;
+            return toSourceTypeName(typeName);
```
[the remark this answers is not in the document]
Comment (novoj) [reply jdwp-mcp-server/src/main/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluator.java:479]: Good catch — fixed in 7c9c9c9. This is reachable in practice: `resolveLocalVarType` passes `var.type()` to `getDeclaredType`, and a local declared as an interface (`List x`, or a package-private interface) yields an `InterfaceType`. A non-public one previously fell through to the raw-name return and broke wrapper compilation. Added an `InterfaceType` branch that DFS-walks `superinterfaces()` for a wrapper-reachable interface (public, or sharing the wrapper package), falling back to `java.lang.Object`, with a visited-guard for the diamond inheritance interfaces routinely exhibit. `isReachableFromWrapper` was generalized from `ClassType` to `ReferenceType` so it serves both kinds. Covered by 5 new tests in `JdiExpressionEvaluatorGetDeclaredTypeTest.InterfaceTypes`.

File: jdwp-mcp-server/src/main/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluator.java [hunk 9]
```diff
@@ -499,9 +699,12 @@ private static ClassLoaderReference findClassLoader(StackFrame frame) throws Jdi
...
      * and returns the result as a JDI {@link Value}. Side effect: populates {@link #compilationCache}.
-     * The auto-rewrite of bare `this.field` references only runs when `this`'s declared type is
-     * public — for non-public types the wrapper class can't reference the type at all, so the
-     * rewrite would just produce a misleading "type not visible" error.
+     *
+     * <p>The auto-rewrite of bare {@code this.field} references runs when EITHER {@code this}'s
+     * declared type is public OR the wrapper is emitted into {@code this}'s own package (see
+     * {@link #resolveWrapperPackage}). In the latter case package-private and protected fields
+     * also become candidates — subject to {@link #isFieldAccessibleFromWrapper}'s per-field
+     * accessibility check.
```
[the remark this answers is not in the document]
Comment (novoj) [reply jdwp-mcp-server/src/main/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluator.java]: Agreed — reworded in 7c9c9c9. Step 2 of the pipeline javadoc now reads: "Auto-rewrite bare field identifiers (`field` -> `_this.field`) via `rewriteThisFieldReferences` when safe; explicit `this`/`this.field` is handled separately by `rewriteThisKeyword`."

File: jdwp-mcp-server/src/main/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluator.java [hunk 10]
```diff
@@ -423,20 +452,229 @@ private static String getDeclaredType(ReferenceType type) {
...
+        final StringBuilder sb = new StringBuilder(n);
+        for (int i = 0; i < n; i++) {
+            final char c = binaryName.charAt(i);
+            if (c == '$' && isNestedClassSeparator(binaryName, i)) {
+                sb.append('.');
+            } else {
+                sb.append(c);
+            }
+        }
+        return sb.toString();
```
[the remark this answers is not in the document]
Comment (novoj) [reply jdwp-mcp-server/src/main/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluator.java:628]: Refuting this one (with a documented limitation rather than a guard). You are right that `Foo$Bar` for a genuine top-level type is misrendered — but `Outer$Inner` (a member type) and a top-level `Foo$Bar` are *identical strings*, so the binary name alone cannot distinguish them. The robust options (compile-then-fallback, or a per-type VM round-trip to confirm the prefix resolves to an enclosing type) are disproportionate here: JLS §3.8 reserves `$` for "mechanically generated source code", so such top-level names do not occur in normal application code (and would be a single pathological type breaking eval at that frame). I added a "Known limitation" paragraph to `toSourceTypeName` documenting exactly this trade-off (2ba7ede). Happy to revisit if a real-world case surfaces.

File: jdwp-mcp-server/src/main/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluator.java [hunk 11]
```diff
@@ -534,24 +775,39 @@ private static ClassLoaderReference findClassLoader(StackFrame frame) throws Jdi
             // NOTE: Classpath configuration must be done BEFORE calling evaluate() to avoid nested JDI calls
             // The caller (e.g., jdwp_evaluate_watchers) is responsible for calling configureCompilerClasspath()
 
+            // 0. Decide which package the wrapper class will live in. For a non-public `this` in an
+            //    addressable package, we emit the wrapper alongside it so package-private fields and
+            //    the type itself become reachable; otherwise we use the default isolated package.
+            final ObjectReference thisObject = frame.thisObject();
+            final String wrapperPackage = resolveWrapperPackage(thisObject);
+
```
[the remark this answers is not in the document]
Comment (novoj) [reply jdwp-mcp-server/src/main/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluator.java:800]: Good catch — fixed in 2ba7ede. The implementation now falls back. I introduced `JdiClassDefinitionException` (a `JdiEvaluationException` subtype) thrown *only* from `RemoteCodeExecutor`'s define phase — before the user method is invoked, so a define failure provably means the expression never ran. `evaluate()` catches it and, when the wrapper package was non-default, retries once in `DEFAULT_EVALUATION_PACKAGE` (rebuilding context, re-running the this.field rewrite, and using a package-qualified cache key). This preserves the pre-feature behaviour for public-only expressions when defining into a sealed/encapsulated app package is rejected. Scoping to the define-phase subtype (rather than any failure) is deliberate: it guarantees a side-effecting expression is never executed twice. Covered by 3 tests (retry / no-retry-when-already-default / no-retry-on-user-code-failure).

File: jdwp-mcp-server/src/test/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluatorGetDeclaredTypeTest.java [hunk 12]
```diff
@@ -143,8 +155,190 @@ void shouldReturnNameForArrayType() throws Exception {
...
+
+		@Test
+		@DisplayName("diamond superinterface graph — visited-guard prevents infinite recursion")
+		void shouldTerminateOnDiamondInheritance() throws Exception {
+			InterfaceType pub = mock(InterfaceType.class);
+			when(pub.name()).thenReturn("com.example.Root");
+			when(pub.isPublic()).thenReturn(true);
+
+			// Two non-public mid-level interfaces both extend the same public Root (a diamond when
+			// a leaf extends both). The visited set must stop Root being re-expanded endlessly.
```
[the remark this answers is not in the document]
Comment (novoj) [reply jdwp-mcp-server/src/test/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluatorGetDeclaredTypeTest.java]: Agreed — reworded in 2ba7ede. Interface inheritance is acyclic, so the DFS terminates regardless of the visited set; the set's real job is to avoid re-expanding a shared parent in a diamond (and to stay safe against pathological/cyclic input). Updated both the test comment and the `findReachableInterface` javadoc to say exactly that.

File: jdwp-mcp-server/src/main/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluator.java [hunk 13]
```diff
@@ -534,106 +788,158 @@ private static ClassLoaderReference findClassLoader(StackFrame frame) throws Jdi
...
+            }
+        } catch (Exception e) {
+            log.warn("[Evaluator] Expression evaluation failed after {}ms: {}",
+                System.currentTimeMillis() - startTime, e.getMessage());
+            // Un-wrap runtime exception from cache computation
+            if (e instanceof RuntimeException && e.getCause() instanceof JdiEvaluationException jdiEx) {
+                throw jdiEx;
+            }
+            throw new JdiEvaluationException("Expression evaluation failed: " + e.getMessage(), e);
+        } finally {
```
[the remark this answers is not in the document]
Comment (novoj) [reply jdwp-mcp-server/src/main/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluator.java:831]: Fixed in 5db2934 — good catch, this was an inconsistency from the last round (I preserved the subtype in RemoteCodeExecutor.execute but then re-wrapped it here). evaluate()'s outer catch now starts with `if (e instanceof JdiEvaluationException jdi) throw jdi;`, so JdiClassDefinitionException (and any other pipeline exception) propagates unchanged with its original message; the RuntimeException-cause unwrap for the cache path is kept after it.

File: jdwp-mcp-server/src/main/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluator.java [hunk 14]
```diff
@@ -18,7 +19,9 @@
  * <p>
  * Pipeline (per call to {@link #evaluate}):
  * 1. Build evaluation context from the frame (locals + `this`).
- * 2. Auto-rewrite bare `this.field` references via {@link #rewriteThisFieldReferences} when safe.
+ * 2. Auto-rewrite bare field identifiers (`field` → `_this.field`) via
+ * {@link #rewriteThisFieldReferences} when safe; explicit `this`/`this.field` is handled
+ * separately by {@link #rewriteThisKeyword}.
  * 3. Cache lookup keyed on `signature + "###" + expression`; on miss, generate a wrapper class
  * with a UUID-suffixed name, compile via {@link InMemoryJavaCompiler}, and cache the bytecode.
```
[the remark this answers is not in the document]
Comment (novoj) [reply jdwp-mcp-server/src/main/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluator.java]: Fixed in 5db2934 — the pipeline javadoc now states the key is `wrapperPackage + "###" + signature + "###" + expression`, with a note that the package is part of the key precisely so a define-time fallback recompiling into a different package gets its own cache entry.

File: jdwp-mcp-server/src/main/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluator.java [hunk 15]
```diff
@@ -499,9 +750,12 @@ private static ClassLoaderReference findClassLoader(StackFrame frame) throws Jdi
...
      * and returns the result as a JDI {@link Value}. Side effect: populates {@link #compilationCache}.
-     * The auto-rewrite of bare `this.field` references only runs when `this`'s declared type is
-     * public — for non-public types the wrapper class can't reference the type at all, so the
-     * rewrite would just produce a misleading "type not visible" error.
+     *
+     * <p>The auto-rewrite of bare {@code this.field} references runs when EITHER {@code this}'s
+     * declared type is public OR the wrapper is emitted into {@code this}'s own package (see
+     * {@link #resolveWrapperPackage}). In the latter case package-private and protected fields
+     * also become candidates — subject to {@link #isFieldAccessibleFromWrapper}'s per-field
+     * accessibility check.
```
[the remark this answers is not in the document]
Comment (novoj) [reply jdwp-mcp-server/src/main/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluator.java]: Fixed in 5db2934. (Step 2 of the class-level pipeline javadoc was already reworded in a prior round; this was a second occurrence in the evaluate() method javadoc.) It now says bare field identifiers (`field` -> `_this.field`) are conditionally rewritten via rewriteThisFieldReferences, while explicit `this`/`this.field` is handled unconditionally by rewriteThisKeyword.

File: jdwp-mcp-server/src/test/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluatorPackageFallbackTest.java [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+ * Covers the define-phase package fallback in {@link JdiExpressionEvaluator#evaluate}: when the
+ * wrapper cannot be DEFINED into a non-public {@code this}'s own package (sealed package / module
+ * encapsulation / restrictive classloader, surfaced as {@link JdiClassDefinitionException}), the
+ * evaluator retries once in the default isolated package. A define-phase failure means the user
+ * expression never ran, so the retry cannot double-execute a side-effecting expression — which is
+ * why only {@link JdiClassDefinitionException} (not a generic user-code failure) triggers it.
+ */
+class JdiExpressionEvaluatorPackageFallbackTest {
+
+	private static final String DEFAULT_PACKAGE = "mcp.jdi.evaluation";
```
[the remark this answers is not in the document]
Comment (novoj) [reply jdwp-mcp-server/src/test/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluatorPackageFallbackTest.java]: Removed in 5db2934 — the unused `DEFAULT_PACKAGE` constant is gone. The tests assert package selection via the `com.app.` vs default prefix in the execute() stub, so the constant was dead.

File: jdwp-mcp-server/src/test/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluatorPackageFallbackTest.java [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+
+	private InMemoryJavaCompiler compiler;
+	private JdiExpressionEvaluator evaluator;
+	private StackFrame frame;
+	private ClassType thisType;
```
[the remark this answers is not in the document]
Comment (novoj) [reply jdwp-mcp-server/src/test/java/one/edee/mcp/jdwp/evaluation/JdiExpressionEvaluatorPackageFallbackTest.java]: Fixed in 5db2934 — `thisType` is now a local inside wireFrame(...) rather than an unread field.
